sycl/include/CL/sycl/detail/program_manager/program_manager.hpp

```diff
@@ -63,26 +63,53 @@ class ProgramManager {
   static string_class getProgramBuildLog(const RT::PiProgram &Program);
 
 private:
-  RT::PiProgram loadProgram(OSModuleHandle M, const context &Context,
-                            DeviceImage **I = nullptr);
-  void build(RT::PiProgram Program, const string_class &Options = "",
-             std::vector<RT::PiDevice> Devices = std::vector<RT::PiDevice>());
-
-  ProgramManager() = default;
+  ProgramManager();
   ~ProgramManager() = default;
   ProgramManager(ProgramManager const &) = delete;
   ProgramManager &operator=(ProgramManager const &) = delete;
 
+  DeviceImage &getDeviceImage(OSModuleHandle M, KernelSetId KSId,
+                              const context &Context);
+  void build(RT::PiProgram Program, const string_class &Options,
+             std::vector<RT::PiDevice> Devices);
+  /// Provides a new kernel set id for grouping kernel names together
+  KernelSetId getNextKernelSetId() const;
+  /// Returns the kernel set associated with the kernel, handles some special
+  /// cases (when reading images from file or using images with no entry info)
+  KernelSetId getKernelSetId(OSModuleHandle M,
+                             const string_class &KernelName) const;
+  /// Returns the format of the binary image
+  RT::PiDeviceBinaryType getFormat(const DeviceImage &Img) const;
+  /// Dumps image to current directory
+  void dumpImage(const DeviceImage &Img, KernelSetId KSId) const;
+
+  using KernelToImgsMap =
+      std::map<KernelSetId, std::unique_ptr<std::vector<DeviceImage *>>>;
+  using StrToKSIdMap = std::map<string_class, KernelSetId>;
```

Comment:
this mapping can be not 1:1 - different images may contain kernels with the same name. The key should apparently be a pair <string_class, OSModuleHandle>. This is is the only problems which I think can't be postponed to a later PR.

Reply:
It is done as a nested map instead of using a pair key right now. So the mapping is still valid.

Reply:
@kbobrovs Added a general comment on kernel resolution